cpp/src/shim/libcurl.cpp

```diff
@@ -116,19 +119,55 @@ CURL* CurlHandle::handle() noexcept { return _handle.get(); }
 
 void CurlHandle::perform()
 {
-  // Perform the curl operation and check for errors.
-  CURLcode err = curl_easy_perform(handle());
-  if (err != CURLE_OK) {
-    std::string msg(_errbuf);  // We can do this because we always initialize `_errbuf` as empty.
-    std::stringstream ss;
-    ss << "curl_easy_perform() error near " << _source_file << ":" << _source_line;
-    if (msg.empty()) {
-      ss << "(" << curl_easy_strerror(err) << ")";
+  long http_code          = 0;
+  int attempt_count       = 1;
+  int base_delay          = 500;   // milliseconds
+  int max_delay           = 4000;  // milliseconds
```

Comment:
We should start from `attempt_count = 0` so that our first delay is 500ms (`500 * (1 << 0) = 500`).
```suggestion
  auto attempt_count        = 0;
  auto constexpr base_delay = 500;   // milliseconds
  auto constexpr max_delay  = 4000;  // milliseconds
```

Reply:
Thanks for catching that. I'll implement this by adjusting the backoff to be `base_delay * (1 << std::min(attempt_count - 1, 4));`, so subtracting 1 there, but keeping the initial attempt at 1. Then the loop logic can still be compared directly against the user-provided max_attempts.